File: src/commands/file/decrypt.ts
import {Command, flags as flagtype} from "@oclif/command";
import {ErrorCodes} from "@ironcorelabs/ironnode";
import * as fs from "fs";
import {extname} from "path";
import chalk from "chalk";
import {ironnode} from "../../lib/SDK";
import {keyFile} from "../../lib/sharedFlags";
import * as Utils from "../../lib/Utils";

type DecryptFileResult = Utils.ErrorOr<{
    decryptedContent: Buffer;
    destinationFileOutput?: string;
}>;

/**
 * File decrypt command. Takes a single or list of files and attempts to decrypt them. If a single file is provided without flags it will attempt to write the
 * decrypted file as a sibling of the encrypted file with it's extension removed. The -o option allows users to have the decrypted content to be written out
 * to either a different file location or `-o -` can be used to print the decrypted content to stdout. If multiple files are provided they will each be written
 * as siblings of their encrypted files with the file extension removed.
 */
export default class Decrypt extends Command {
    static description = "Decrypt a file or list of files. By default, the decrypted file is written to the same directory without the '.iron' extension.";
    static strict = false; //Allow multiple files to be provided as arguments
    static args = [
        {
            name: "files",
            description: "Path of file or files to decrypt.",
        },
    ];
    static flags = {
        help: flagtype.help({char: "h"}),
        keyfile: keyFile(),
        out: flagtype.string({
            char: "o",
            description:
                "Filename where decrypted file will be written. Only allowed if a single file is being decrypted. Use '-o -' to write decrypted file content to stdout.",
        }),
        stdin: flagtype.boolean({
            char: "s",
            description: "Read data to decrypt from stdin. If used, no source files should be provided as arguments and you must use the -o flag.",
            dependsOn: ["out"],
        }),
        delete: flagtype.boolean({
            char: "d",
            description: "Delete the input file after successful decrypt",
            exclusive: ["stdin"],
        }),
    };
    static examples = [
        Utils.buildCommandSampleText(
            "file:decrypt path/to/file.txt.iron",
            "Decrypt a file given its path and write it out to 'path/to/file.txt' leaving input file in place."
        ),
        Utils.buildCommandSampleText(
            "file:decrypt path/to/file.json.iron -o file.json",
            "Decrypt the provided encrypted file and write the result to file.json in the current working directory."
        ),
        Utils.buildCommandSampleText(
            "file:decrypt path/to/file.json.iron -d -o file.json",
            "Decrypt the provided encrypted file and write the result to file.json and delete the encrypted path/to/file.json.iron file."
        ),
        Utils.buildCommandSampleText("file:decrypt *.iron", "Decrypt all of the '.iron' files and write them to files without the '.iron' extension."),
        Utils.buildCommandSampleText("file:decrypt path/to/file.iron -o -", "Decrypt the provided file and write the decrypted bytes to stdout."),
        Utils.buildCommandSampleText(
            `${chalk.green("cat")} encryptedfile.iron | ${chalk.green("ironhide")} file:decrypt -s -o -`,
            "Decrypt the provided file from stdin and write the decrypted bytes to stdout.",
            false
        ),
    ];

    /**
     * Strip off the last extension from the provided file path. Would remove .iron from both file.json.iron and file.iron
     */
    removeLastFileExtension(filePath: string) {
        const lastExtensionLength = extname(filePath).length;
        if (lastExtensionLength === 0) {
            //The file we're trying to decrypt doesn't have an extension, so just return the file name. This is very likely
            //to fail later on because we're trying to write the decrypted content to the same place as the encrypted file.
            return filePath;
        }
        return filePath.slice(0, -lastExtensionLength);
    }

    /**
     * Read all the data from stdin into a single Buffer since we don't yet support streaming decryption. Returns a Promise which
     * resolves with all file data as bytes or rejects with any error encountered while reading the stream.
     */
    readAllContentFromStdin(): Promise<Buffer> {
        const encryptedData: Buffer[] = [];
        return new Promise((resolve, reject) => {
            process.stdin.on("data", (chunk: Buffer) => encryptedData.push(chunk));
            process.stdin.on("end", () => resolve(Buffer.concat(encryptedData)));
            process.stdin.on("error", (e) => reject(e));
        });
    }

    /**
     * Returns the bytes to decrypt conditionally based off whether the user is decrypting a file provided on the CLI or from content
     * read from stdin.
     */
    getBytesToDecrypt(decryptOptions: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp): Promise<Buffer> {
        if (Utils.isStdInFileOperation(decryptOptions)) {
            return this.readAllContentFromStdin();
        }
        const sourceFile = Utils.normalizePathToFile(decryptOptions.file);
        return Utils.checkSourceFilePermissions(sourceFile).then(() => fs.readFileSync(sourceFile));
    }

    /**
     * If provided, verify the provided output file. If not provided, just resolves.
     */
    verifyOutfile(outPath?: string) {
        return outPath ? Utils.checkDestinationFilePermissions(Utils.normalizePathToFile(outPath)) : Promise.resolve();
    }

    /**
     * Given an encrypted file path, read in the file content to bytes then extract the document ID from the doc
     * and use that to decrypt the document. Returns the decrypted document response from the SDK
     */
    getDocumentIDAndDecrypt(encryptedBytes: Buffer) {
        return ironnode()
            .document.getDocumentIDFromBytes(encryptedBytes)
            .then((documentID) => {
                if (!documentID) {
                    throw new Error(`Could not parse encrypted file.`);
                }
                return ironnode().document.decryptBytes(documentID, encryptedBytes);
            })
            .catch((e) => {
                throw new Error(e.code === ErrorCodes.DOCUMENT_HEADER_PARSE_FAILURE ? `Failed to decrypt. Input doesn't appear to be encrypted.` : e.message);
            });
    }

    /**
     * Returns a Promise of decrypting a file. Checks the file permissions of the provided input and output paths and attempts
     * to decrypt the file and write the result to the output file. If no output file is provided we take the same path as the
     * input file and strip off it's last extension.
     */
    getFileDecryptPromise(decryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp): Promise<DecryptFileResult> {
        let destinationFileOutput: string | undefined;
        // tslint:disable-next-line
        if (decryptOp.out) {
            destinationFileOutput = decryptOp.out === "-" ? undefined : decryptOp.out;
        } else {
            destinationFileOutput = this.removeLastFileExtension((decryptOp as Utils.ProcessSingleFileOp).file);
        }

        return this.verifyOutfile(destinationFileOutput)
            .then(() => this.getBytesToDecrypt(decryptOp))
            .then((encryptedBytes) => this.getDocumentIDAndDecrypt(encryptedBytes))
            .then((decryptedDocument) => {
                if (!Utils.isStdInFileOperation(decryptOp) && decryptOp.deleteSource) {
                    try {
                        Utils.deleteFile(decryptOp.file);
                    } catch (_) {
                        this.warn(`Unable to delete encrypted source file '${decryptOp.file}' as it is not writable.`);
                    }
                }
                return {
                    decryptedContent: decryptedDocument.data,
                    destinationFileOutput,
                };
            })
            .catch((e) => e);
    }

    /**
     * Decrypt a collection of files and write them out to the same location but with their extension removed.
     */
    decryptMultipleFiles(decryptOp: Utils.ProcessMultipleFileOp) {
        const decryptOps = decryptOp.file.map((file) => this.getFileDecryptPromise({file, out: decryptOp.out, deleteSource: decryptOp.deleteSource}));

        return Promise.all(decryptOps).then((results) => {
            let successfulCount = 0;
            let failureCount = 0;
            results.forEach((decryptResult) => {
                if (Utils.isError(decryptResult)) {
                    this.log(chalk.red(decryptResult.message));
                    failureCount++;
                } else {
                    fs.writeFileSync(decryptResult.destinationFileOutput as string, decryptResult.decryptedContent);
                    successfulCount++;
                }
            });
            if (successfulCount > 0) {
                this.log(chalk.green(`\n${successfulCount} file(s) successfully decrypted.`));
            }
            if (failureCount > 0) {
                this.log(chalk.red(`\n${failureCount} file(s) failed to be decrypted.`));
            }
        });
    }

    /**
     * Decrypt a single file and write the decrypted bytes either to a sibling file, the location requested by the -o option or
     * write the bytes to stdout.
     */
    decryptSingleFile(decryptOp: Utils.ProcessSingleFileOp | Utils.ProcessStdinOp) {
        return this.getFileDecryptPromise(decryptOp).then((result) => {
            if (Utils.isError(result)) {
                throw result;
            }
            if (decryptOp.out === "-") {
                //Write out the decrypted bytes to stdout
                process.stdout.write(result.decryptedContent);
            } else {
                //Write out the decrypted bytes to the file and display a success message
                fs.writeFileSync(result.destinationFileOutput as string, result.decryptedContent);
                const writtenResult = Utils.isStdInFileOperation(decryptOp) ? decryptOp.out : this.removeLastFileExtension(decryptOp.file);
                this.log(chalk.green(`File successfully decrypted and written to ${writtenResult}.`));
            }
        });
    }

    /**
     * Manual validation of complex flag and argument combinations to make sure the user is doing the right thing. Also limit the number of files that
     * we attempt to decrypt at any one time.
     */
    checkFlagsAndArguments(argv: string[], stdin: boolean, out?: string) {
        if (argv.length > 1 && out) {
            return this.error(chalk.red("The output flag (-o) can only be provided when decrypting a single file."));
        }
        if (argv.length === 0 && !stdin) {
            return this.error(chalk.red("Error: You must either provide a path to a file to decrypt or use the -s flag to read from stdin."));
        }
        if (argv.length > 0 && stdin) {
            return this.error(chalk.red("Error: You cannot provide both source files and the -s flag to read from stdin."));
        }
    }

    async run() {
        const {argv, args, flags} = this.parse(Decrypt);
        this.checkFlagsAndArguments(argv, flags.stdin, flags.out);
        let decryptOp: Utils.ProcessOp;
        if (flags.stdin) {
            decryptOp = {out: flags.out as string} as Utils.ProcessStdinOp;
        } else {
            decryptOp = {
                file: argv.length > 1 ? argv : (args.files as string),
                out: flags.out,
                deleteSource: flags.delete,
            } as Utils.ProcessFileOp;
        }

        try {
            Utils.isMultipleFileOperation(decryptOp) ? await this.decryptMultipleFiles(decryptOp) : await this.decryptSingleFile(decryptOp);
        } catch (e) {
            this.error(chalk.red(e.message));
        }
    }
}

File: src/file/decrypt.rs
use crate::util::{self, act_on_all_files};
use clap::Parser;
use ironoxide::prelude::BlockingIronOxide;
use std::{
    fs::{self, File, OpenOptions},
    io::{self, Read, Write},
    path::Path,
    path::PathBuf,
};
use yansi::Paint;

const EXAMPLES: &str = "EXAMPLES

    Decrypt a file given its path and write it out to 'path/to/file.txt' leaving input file in place.
        $ ironhide file decrypt path/to/file.txt.iron

    Decrypt the provided encrypted file and write the result to file.json in the current working directory.
        $ ironhide file decrypt path/to/file.json.iron -o file.json

    Decrypt the provided encrypted file and write the result to file.json and delete the encrypted
    path/to/file.json.iron file.
        $ ironhide file decrypt path/to/file.json.iron -d -o file.json

    Decrypt all of the '.iron' files and write them to files without the '.iron' extension.
        $ ironhide file decrypt *.iron

    Decrypt the provided file and write the decrypted bytes to stdout.
        $ ironhide file decrypt path/to/file.iron -o -

    Decrypt the provided file from stdin and write the decrypted bytes to stdout.
        $  cat encryptedfile.iron | ironhide file decrypt -s -o -
";

#[derive(Parser)]
#[clap(after_help = EXAMPLES)]
/// Decrypt a file or list of files. By default, the decrypted file is written to the same directory without the
/// '.iron' extension.
pub struct Decrypt {
    /// Delete the encrypted source file(s) after successful encryption.
    #[clap(short, long, takes_value = false)]
    delete: bool,
    /// Path of file or files to decrypt.
    #[clap(parse(from_os_str), min_values = 1, required = true)]
    files: Vec<PathBuf>,
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(parse(from_os_str), short, long)]
    keyfile: Option<PathBuf>,
    /// Filename where decrypted file will be written. Only allowed if a single file is
    /// being decrypted.
    /// Use '-o -' to write decrypted file content to stdout, but fair warning, the output is binary and not ASCII.
    #[clap(parse(from_os_str), short, long)]
    out: Option<PathBuf>,
    /// Read data to decrypt from stdin. If used, no source files should be provided as
    /// arguments and you must use the '-o' flag.
    #[clap(
        short,
        long,
        takes_value = false,
        conflicts_with = "files",
        requires = "out"
    )]
    stdin: bool,
}

impl util::GetKeyfile for Decrypt {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

pub fn decrypt_files(
    sdk: &BlockingIronOxide,
    Decrypt {
        delete,
        files,
        out,
        stdin,
        ..
    }: Decrypt,
) -> Result<(), String> {
    if stdin {
        let mut encrypted_document: Vec<u8> = vec![];
        match io::stdin().read_to_end(&mut encrypted_document) {
            Ok(_) => {
                // safe to unwrap here because `stdin` always has to have `out` set.
                let out_path = out.unwrap();
                decrypt_file(sdk, encrypted_document, None, out_path, delete)?;
            }
            Err(e) => util::println_paint(Paint::red(format!("Error reading stdin: {}", e))),
        }
    } else if out.is_some() && files.len() > 1 {
        util::println_paint(Paint::red(
            "Cannot use '-o' flag with multiple files.".to_string(),
        ));
    } else {
        act_on_all_files(
            &files,
            |path: &PathBuf| -> Result<(), String> {
                let mut file = File::open(&path).map_err(|e| {
                    format!(
                        "Provided path '{}' doesn't exist or is not readable: {e}",
                        path.display()
                    )
                })?;
                let mut encrypted_document = Vec::new();
                file.read_to_end(&mut encrypted_document).map_err(|e| {
                    format!(
                        "Failed to read bytes from the encrypted document at {}: {e}",
                        path.display()
                    )
                })?;
                let in_parent = path.parent().ok_or(format!(
                    "Failed to find parent of input path {}.",
                    path.display()
                ))?;
                let out_path =
                    out.clone()
                        .unwrap_or(in_parent.join(Path::new(path.file_stem().ok_or(format!(
                            "Failed to extract default output file name from input path {}.",
                            path.display()
                        ))?)));
                decrypt_file(
                    sdk,
                    encrypted_document,
                    Some(path),
                    out_path.clone(),
                    delete,
                )?;
                if files.len() == 1 {
                    let out_logged_path = get_output_logged_path(out_path)?;
                    util::println_paint(Paint::green(format!(
                        "File successfully decrypted and written to {}",
                        out_logged_path
                    )));
                }
                Ok(())
            },
            "decrypted",
        )?;
    }

    Ok(())
}

fn get_output_writer(out_path: PathBuf) -> Result<Box<dyn Write>, String> {
    let out_writer: Box<dyn Write> = if out_path == PathBuf::from("-") {
        Box::new(io::stdout())
    } else {
        OpenOptions::new()
            .create_new(true)
            .write(true)
            .append(true)
            .open(out_path.clone())
            .map_err(|e| {
                format!(
                    "Couldn't create a file at the desired output path '{}': {e}",
                    out_path.display()
                )
            })
            .map(Box::new)?
    };
    Ok(out_writer)
}

fn get_output_logged_path(out_path: PathBuf) -> Result<String, String> {
    let out_logged_path = if out_path == PathBuf::from("-") {
        "stdout"
    } else {
        out_path.to_str().ok_or(format!(
            "Output path '{}' wasn't valid unicode.",
            out_path.display()
        ))?
    };
    Ok(out_logged_path.to_string())
}

fn decrypt_file(
    sdk: &BlockingIronOxide,
    encrypted_document: Vec<u8>,
    input_path: Option<&PathBuf>,
    out_path: PathBuf,
    delete: bool,
) -> Result<(), String> {
    let decrypt_result = sdk
        .document_decrypt(&encrypted_document)
        .map_err(|e| format!("Failed to decrypt encrypted document: {e}"))?;
    let decrypted_document = decrypt_result.decrypted_data();
    let mut decrypted_writer = get_output_writer(out_path)?;
    decrypted_writer
        .write_all(decrypted_document)
        .map_err(|e| format!("Failed to write decrypted document: {e}"))?;
    if delete {
        match input_path {
            Some(infile) => {
                if fs::remove_file(&infile).is_err() {
                    util::println_paint(Paint::yellow(format!(
                        "Unable to delete source file '{}' as it is not writable.",
                        &infile.display()
                    )))
                }
            }
            None => util::println_paint(Paint::yellow(
                "Unable to delete source as it was a stream.".to_string(),
            )),
        }
    }
    Ok(())
}
